### PR TITLE
feat: improve plotting performance

### PR DESCRIPTION
**Pull request by maehler**
This PR aims to improve the plotting performance in the CNV report. The way that this has been done is to use a sliding window approach where the windows are based on pixels which are then translated into genomic coordinates. This way the windows will be proportional to the view, regardless of how big the sequence is, and it makes it easier to control the resolution of the data. At a certain zoom level (currently when there are windows smaller than 4 bp) all the original data will be displayed.

What I've been struggling a bit with is how to represent each window in the plots. So far I've considered three strategies.

1. Within each window, a random point gets picked and represented (this is what is currently implemented on this branch). Here is where I'm struggling a bit. The rationale behind picking a random point instead of the mean or median is that this would be more likely to sample a point where the data is most dense. Another reason is that in the VAF plot there are windows where values are mirrored around 0.5, and using a summary value would completely hide this. Of course, there could be different versions of the function for log<sub>2</sub> values and VAF, but I would like to avoid this. A downside of this approach is that extreme values might be missed if they are not well represented within a window.
Example:
image

2. Another approach that I tried was to show the maximum and minimum value within each window, but with this it looked like the data had a bimodal distribution, which in itself is a bit misleading. On the other hand, this is able to represent the extreme values in each window. Another thing that becomes misleading is the VAF plot where segments that have a real VAF around 0.5 suddenly look like it is mirrored around 0.5.
Example:
image

3. Yet another approach I've been thinking of, but not tested, is to represent each window as a point (e.g. median) with whiskers that would represent the extreme values within that window. Something I like about this is that you get an immediate visual cue that you are looking at a summary of the data. When zooming in, the whiskers would disappear, indicating that you see the raw data. The downside of this is that it adds some extra complexity to the plotting, both visually and implementation-wise. It might also affect performance negatively, but likely much less than plotting all data points at all times.

For reference, here is the how the plot would look with all the points plotted:
image

Note that the plots are based on simulated random data and have no biological origin whatsoever.

The reason this is marked as a feature and not a performance fix is that the view will not be identical to previous versions given the same data.

### CODE REVIEW

**Comment by maehler** [comment]
Here is a new version where each window is represented as the mean &plusmn; 1 standard deviation.

image

When zoomed in enough all points are shown instead. As you can see, in cases where a window overlaps a breakpoint, the standard deviation becomes quite high. but zooming in to these regions will reveal what is actually going on.

For the log<sub>2</sub> ratios the calculation is straightforward, but in order to capture the bimodal nature of the allele frequencies I made some assumptions. I first bin all points within a window into five bins between 0 and 1. If most points are found in the third bin, i.e. centered around 0, I assume that the distribution is not bimodal and only calculate one mean. If most values are found in any other bin, I assume that the distribution is bimodal. Since it should be centered around 0.5 I calculate two values for these windows; one for all values &ge; 0.5 and one for all values &lt; 0.5. This seems to work well for the dummy data I have, but I will also test this on real data before proceeding.

**Comment by maehler** [comment]
I've tested it on some real data too, and based on that I made some adjustments. It probably needs more tweaking in order to cover different scenarios, but this is a start. Let me know what you think!

**Comment by jonca79** [comment]
I really like the mean +/- std dev! Nice! For the copy numbers I think it will perform very nice. For the VAF-values I'm a bit concerned for the low TC cases (20%) where all values will probably land in the middle bin even when there is a deletion or a duplication. However, it might be fine anyway as the height of the error bars should increase in these areas and the del / dup might be more visible in the summary data than the noisy raw data. 
It might look a bit strange though in cases when the data flips between bimodal and monomodal (is that a word?) for different windows when data is on the limit between the two.

**Comment by maehler** [comment]
I see your point, and I think it's a valid one. At the same time it is difficult to come up with a universal solution without having seen more examples. At the moment there has to be at over 300 points in the VAF panel for them to be summarise, otherwise all points are shown. Whether or not this is a reasonable threshold I simply don't know from the little testing that I've done.

I will run all our validation samples through this and see how it looks for them and do some tweaking based on the results.

**Comment by maehler** [comment]
Now I finally managed to free some time to run through some samples for testing, and as I suspected, not all aspects of my implementation work nicely for real data. The main problem seems to be sparsity, especially when it comes to the VAFs. Here I would have to play around with thresholds and what the smallest unit of plotting should be. I will try to spend some time here and see if I can come up with something that would be reasonably general across the datasets that I have.

**Comment by maehler** [comment]
Now the data in the genome view is summarised or not based on the average number of data points per chromosome. If the average value is over the currently hard-coded threshold (currently 300 points) the data is summarised, otherwise not. This is done independently on the log<sub>2</sub> ratios and the VAFs.

In order to make it easier to see what is going on, I added a checkbox that allows the user to force the display of all data points in the chromosome view. Below is an example from a HD832 sample.

cnv_data_summary

Especially the VAFs look messy in this example. In the patient samples I've been looking at, none of them have had enough data for the VAFs to be summarised with the current threshold, and they in general look much more tidy. I tried forcing the data to be summarised for these samples, and the separation in the VAFs was clearly visible where this effect was apparent. In the less apparent cases, the standard deviation in the summarised data gave a hint that something was going on underneath, and with the possibility to toggle between summarised and raw data I think should be adequate for exploring.

Let me know what you think.

File: workflow/scripts/cnv_html_report.py
Comment (maehler) [inline workflow/scripts/cnv_html_report.py:1]: This actually fixes a bug I missed in the merge of #48.

**Comment by maehler** [comment]
Segemnt colour before last commit:
image
Segment colour after last commit:
image

**Review by jonca79 — APPROVED** [review]

**Pull request merged**